compiler/transforms/convert_acc_to_csr.py

```diff
@@ -0,0 +1,273 @@
+from abc import ABC
+from dataclasses import dataclass
+from functools import cache
+
+from xdsl.dialects import arith, builtin, llvm
+from xdsl.dialects.builtin import StringAttr, i32
+from xdsl.dialects.scf import Condition, While, Yield
+from xdsl.ir import MLContext, Operation
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.traits import SymbolTable
+
+from compiler.dialects import acc
+
+
+@dataclass
+class LowerAccBasePattern(RewritePattern, ABC):
+    """
+    Base class for the acc2 dialect lowerings.
+
+    Wraps some common logic to get handles to accelerator ops inside the module.
+    """
+
+    module: builtin.ModuleOp
+
+    @cache
+    def get_acc(self, accelerator: StringAttr) -> acc.AcceleratorOp:
+        """
+        Get a reference to the accelerator
+        """
+        trait = self.module.get_trait(SymbolTable)
+        assert trait is not None
+        acc_op = trait.lookup_symbol(self.module, accelerator)
+        if not isinstance(acc_op, acc.AcceleratorOp):
+            raise RuntimeError(
+                f"Invalid IR: no accelerator op for @{accelerator.data} found in module"
+            )
+        return acc_op
+
+    def __hash__(self):
+        return id(self)
+
+
+class LowerAccSetupToCsr(LowerAccBasePattern):
+    """
+    Convert setup ops to a series of CSR sets that set each field to the given value.
+
+    Looks up the csr addresses of the value fields by getting the `acc2.accelerator`
+    operation from the module op.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: acc.SetupOp, rewriter: PatternRewriter, /):
+        # grab a dict that translates field names to CSR addresses:
+        field_to_csr = dict(self.get_acc(op.accelerator).field_items())
+
+        # emit the llvm assembly code to set csr values:
+        for field, val in op.iter_params():
+            addr = field_to_csr[field]
+            rewriter.insert_op_before_matched_op(
+                [
+                    addr_val := arith.Constant(addr),
+                    llvm.InlineAsmOp(
+                        "csrw $0, $1",
+                        "I, r",
+                        [addr_val, val],
+                        has_side_effects=True,
```

Comment:
Can we do the "I, K" constraint here too? Or is llvm not smart enough to decide which is best?

Reply:
LLVM is smart!